core/lib/zksync_core/src/state_keeper/io/fee_address_migration.rs

```diff
@@ -0,0 +1,304 @@
+//! Temporary module for migrating fee addresses from L1 batches to miniblocks.
+
+use std::{
+    ops,
+    time::{Duration, Instant},
+};
+
+use anyhow::Context as _;
+use async_trait::async_trait;
+use tokio::sync::watch;
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_types::{Address, MiniblockNumber};
+
+use crate::utils::{binary_search_with, BinarySearchPredicate};
+
+/// Runs the migration for pending miniblocks.
+pub(crate) async fn migrate_pending_miniblocks(storage: &mut StorageProcessor<'_>) {
+    let started_at = Instant::now();
+    tracing::info!("Started migrating `fee_account_address` for pending miniblocks");
+
+    // FIXME: should work after 2nd DB migration (currently, will panic)
+    let rows_affected = storage
+        .blocks_dal()
+        .copy_fee_account_address_for_pending_miniblocks()
+        .await
+        .expect("Failed migrating `fee_account_address` for pending miniblocks");
+    let elapsed = started_at.elapsed();
+    tracing::info!("Migrated `fee_account_address` for {rows_affected} miniblocks in {elapsed:?}");
+}
+
+/// Runs the migration for non-pending miniblocks. Should be run as a background task.
+pub(crate) async fn migrate_miniblocks(
+    pool: ConnectionPool,
+    last_miniblock: MiniblockNumber,
+    stop_receiver: watch::Receiver<bool>,
+) -> anyhow::Result<()> {
+    let MigrationOutput {
+        migrated_range,
+        miniblocks_affected,
+    } = migrate_miniblocks_inner(
+        pool,
+        last_miniblock,
+        100_000,
+        Duration::from_secs(1),
+        stop_receiver,
+    )
+    .await?;
+
+    tracing::info!(
+        "Finished fee address migration for range {migrated_range:?} with {miniblocks_affected} affected miniblocks"
+    );
+    Ok(())
+}
+
+#[derive(Debug, Default)]
+struct MigrationOutput {
+    migrated_range: Option<ops::RangeInclusive<MiniblockNumber>>,
+    miniblocks_affected: u64,
+}
+
+async fn migrate_miniblocks_inner(
+    pool: ConnectionPool,
+    last_miniblock: MiniblockNumber,
+    chunk_size: u32,
+    sleep_interval: Duration,
+    stop_receiver: watch::Receiver<bool>,
+) -> anyhow::Result<MigrationOutput> {
+    anyhow::ensure!(chunk_size > 0, "Chunk size must be positive");
+
+    let mut storage = pool.access_storage().await?;
+    // Check whether the last miniblock is migrated. If it is, we can finish early.
+    // This will also protect us from running unsupported `BlocksDal` queries after the 2nd DB migration is run
+    // removing `l1_batches.fee_account_address`.
+    if is_fee_address_migrated(&mut storage, last_miniblock).await? {
```

Comment:
It seems that your approach assumes that migration is run in a one-shot. If the process was terminated and there were some non-migrated miniblocks then they won't be migrated on restart. I mean the following scenario:
You launch server, it migrates pending miniblocks and starts migration for others -> meanwhile some batch is sealed so some blocks are not considered to be pending anymore -> process exits -> on restart  `is_fee_address_migrated(&mut storage, last_miniblock).await` returns true, so migration is skipped.

As a solution, you can create some temp DB table and save there last sealed miniblock number when you're calling `copy_fee_account_address_for_pending_miniblocks` for the first time. Then pass this number to `migrate_miniblocks` instead of `self.io.current_miniblock_number() - 1`. Having a temporary table looks pretty ugly but I don't have other solutions in mind

Reply:
Yep, that's a problem. I've spent some effort to make the migration fault-tolerant (there's even a corresponding test), but it evidently wasn't enough 🙃

In terms of alternatives: What about checking the start element of each chunk of miniblocks to see whether it needs to be migrated instead of using binary search? As long as the chunk size is a hard-coded constant, it seems to work, and if the chunk size is large, it'll be reasonably efficient (e.g., if the chunk size is 100,000, we'd need to check approx. 220 addresses on the mainnet, and we could do this in a single DB query). A failure case would be if we deem that the chunk is migrated, but it's really not (or at least not *fully* migrated). IIUC, this can only happen if we assume that the node is downgraded and produces some miniblocks w/o `fee_account_address` after the migration is complete (but in this case, your approach won't work either).

Reply:
Your alternative looks better, I think it's ok to forbid downgrading